src/compiler/checker.ts

```diff
@@ -33675,15 +33675,32 @@ namespace ts {
                 else if (isChildOfNodeWithKind(node, SyntaxKind.EnumMember)) {
                     diagnosticMessage = Diagnostics.Octal_literals_are_not_allowed_in_enums_members_initializer_Use_the_syntax_0;
                 }
-                if (diagnosticMessage) {
-                    const withMinus = isPrefixUnaryExpression(node.parent) && node.parent.operator === SyntaxKind.MinusToken;
-                    const literal = (withMinus ? "-" : "") + "0o" + node.text;
-                    return grammarErrorOnNode(withMinus ? node.parent : node, diagnosticMessage, literal);
-                }
             }
+
+            if (!diagnosticMessage && numericLiteralValueImpreciselyLarge(node.text)) {
+                diagnosticMessage = Diagnostics.Numeric_literals_with_absolute_values_equal_to_2_53_or_greater_are_too_large_to_be_represented_accurately_as_an_integer;
+            }
+
+            if (diagnosticMessage) {
+                const withMinus = isPrefixUnaryExpression(node.parent) && node.parent.operator === SyntaxKind.MinusToken;
+                const literal = (withMinus ? "-" : "") + "0o" + node.text;
+                return grammarErrorOnNode(withMinus ? node.parent : node, diagnosticMessage, literal);
+            }
+
             return false;
         }
 
+        function numericLiteralValueImpreciselyLarge(text: string) {
+            // We can quickly bail out in two common cases:
+            // * Literals with 15 or fewer characters, as they aren't long enough to reach 2^53 - 1
+            // * Fractional numbers (e.g. 9000000000000000000000000000.000000001) are inherently imprecise anyway
+            if (text.length <= 15 || text.indexOf(".") !== -1) {
+                return false;
+            }
+
+            return Number(text) >= Math.pow(2, 53) - 1;

```

Comment:
Hmm, per the user baselines, should allow `9007199254740991` as  `2**53`. Will update...

```suggestion
            return Number(text) >= 2 ** 53;
```

Reply:
IE does not have the exponentiation operator, according to mdn

Reply:
it gets downleveled

Reply:
Right. I forgot TS down levels syntax, but not methods.